streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -694,61 +571,58 @@ private void closeTopology() {
             closeStateManager(clean);
-        } catch (final RuntimeException e) {
-            if (firstException == null) {
-                firstException = e;
+        } catch (final RuntimeException error) {
+            if (clean) {
+                throw error;
             }
-            log.error("Could not close state manager due to the following error:", e);
-        }
-
-        partitionGroup.close();
-        streamsMetrics.removeAllTaskLevelSensors(threadId, id.toString());
-
-        closeTaskSensor.record();
+        } finally {
+            partitionGroup.close();
+            recordCollector.close();
+            closeTaskSensor.record();
 
-        if (firstException != null) {
-            throw firstException;
+            streamsMetrics.removeAllTaskLevelSensors(threadId, id.toString());
         }
     }
 
     /**
      * <pre>
-     * - {@link #suspend(boolean, boolean) suspend(clean)}
-     *   - close topology
-     *   - if (clean) {@link #commit()}
-     *     - flush state and producer
-     *     - commit offsets
-     * - close state
-     *   - if (clean) write checkpoint
-     * - if (eos) close producer
+     * the following order must be followed:
+     *  1. first close topology to make sure all cached records in the topology are processed
+     *  2. then flush the state, send any left changelog records
+     *  3. then flush the record collector
+     *  4. then commit the record collector -- for EOS this is the synchronization barrier
+     *  5. then checkpoint the state manager -- even if we crash before this step, EOS is still guaranteed
      * </pre>
      *
      * @param clean    shut down cleanly (ie, incl. flush and commit) if {@code true} --
      *                 otherwise, just close open resources
-     * @param isZombie {@code true} is this task is a zombie or not (this will repress {@link TaskMigratedException}
      * @throws TaskMigratedException if committing offsets failed (non-EOS)
      *                               or if the task producer got fenced (EOS)
      */
     @Override
-    public void close(boolean clean,
-                      final boolean isZombie) {
+    public void close(final boolean clean) {
         log.debug("Closing");
 
-        RuntimeException firstException = null;
         try {
-            suspend(clean, isZombie);
-        } catch (final RuntimeException e) {
-            clean = false;
-            firstException = e;
-            log.error("Could not close task due to the following error:", e);
+            // If it is an unclean close, then the suspend would never
+            // throw since it would swallow all exception; otherwise if suspension
+            // throws we do not need to proceed to further steps and should just notify
+            // the caller thread. Therefore it is always safe to proceed without try-catch
+            suspend(clean);
+
+            closeSuspended(clean);
+        } catch (final RuntimeException error) {
```

Comment:
Yes, sounds good to me!

Reply:
One tricky thing is that in commit / suspend there are several steps to execute, and when `clean = false` we should still execute all the steps even if early steps throw.

Reply:
Didn't follow this last thought. If we have a pattern where methods with `clean=false` never throw (which was my suggestion), it would automatically do what you are saying here, right?

Reply:
There are two things with `clean=false` here: 1) we should not throw any more, and 2) we can skip some steps like trying to commit, because we know it is likely to fail, and even it does not it's not meaningful anyways. 2) is an optional optimization admittedly.